src/iexamine.cpp

```diff
@@ -2674,6 +2675,104 @@ void iexamine::shrub_wildveggies( player &p, const tripoint &examp )
     return;
 }
 
+void iexamine::recycle_compactor(player &, const tripoint &examp)
+{
+    // choose what metal to recycle
+    auto metals = materials::get_compactable();
+    uimenu choose_metal;
+    choose_metal.text = _("Recycle what metal?");
+    for( auto &m : metals ) {
+        choose_metal.addentry(m.name());
+    }
+    choose_metal.entries.push_back( uimenu_entry( -1, true, 'c', _("Cancel") ) );
+    choose_metal.selected = choose_metal.entries.size();
+    choose_metal.query();
+    int m_idx = choose_metal.ret;
+    if( m_idx < 0 || m_idx >= static_cast<int>(metals.size()) ) {
+        add_msg(_("Never mind."));
+        return;
+    }
+    material_type m = metals.at(m_idx);
+
+    // check inputs and tally total mass
+    auto inputs = g->m.i_at(examp);
+    units::mass sum_weight = 0;
+    for( auto it = inputs.begin(); it != inputs.end(); ++it ) {
+        if( !it->only_made_of( std::set<material_id> { m.id } ) ) {
+            add_msg(_("You realize this isn't going to work because %s is not made purely of %s."), it->tname().c_str(), m.name().c_str());
```

Comment:
Please use positional arguments here and below. Consider also adding translator comments.